Tensile/Source/client/source/HardwareMonitor.cpp

```diff
@@ -386,9 +394,184 @@ namespace Tensile
+                    return std::numeric_limits<double>::quiet_NaN();
+                    
+                return (std::accumulate(m_freqValues.begin(), m_freqValues.end(), double(0))) / m_freqValues.size() ;
+            }
+
+            throw std::runtime_error("Can't read Gfx value that wasn't requested: ");
+        }
+        
+        double HardwareMonitor::getMedianPower()
+        {
+            assertNotActive();
+          
+            if(m_dataPoints == 0)
+                throw std::runtime_error("No data points collected!");
+                
+            if(m_powerValues.size() != 0)
+            {
+                if(m_powerValues[0] == std::numeric_limits<uint16_t>::max())
+                    return std::numeric_limits<double>::quiet_NaN();
+
+                size_t powerMidValueIndex = m_powerValues.size() / 2;
+                std::nth_element(m_powerValues.begin(), m_powerValues.begin()+powerMidValueIndex, m_powerValues.end());
+                return static_cast<double> (m_powerValues[powerMidValueIndex]);
+            }
+           
+            throw std::runtime_error("Can't read Gfx value that wasn't requested: ");
+        }
+        
+        double HardwareMonitor::getAveragePower()
+        {
+            assertNotActive();
+
+            if(m_dataPoints == 0)
+                throw std::runtime_error("No data points collected!");
+
+            if(m_powerValues.size() != 0)
+            {
+                if(m_powerValues[0] == std::numeric_limits<uint16_t>::max())
+                    return std::numeric_limits<double>::quiet_NaN();
+                    
+                return (std::accumulate(m_powerValues.begin(), m_powerValues.end(), double(0))) 
+                                                             / m_powerValues.size() ;
+            }
+            
+            throw std::runtime_error("Can't read Power value that wasn't requested: ");
+        }
+        
+        double HardwareMonitor::getMedianHotSpotTemperature()
+        {
+            assertNotActive();
+            
+            if(m_dataPoints == 0)
+                throw std::runtime_error("No data points collected!");
+
+            if(m_tempHotspotValues.size() != 0)
+            {
+                if(m_tempHotspotValues[0] == std::numeric_limits<uint16_t>::max())
+                    return std::numeric_limits<double>::quiet_NaN();
+                    
+                size_t temperatureMidValueIndex = m_tempHotspotValues.size() / 2;
+                std::nth_element(m_tempHotspotValues.begin(), m_tempHotspotValues.begin()+temperatureMidValueIndex, m_tempHotspotValues.end());
+                return static_cast<double> (m_tempHotspotValues[temperatureMidValueIndex]);
+            }
+            
+            throw std::runtime_error("Can't read temperature value that wasn't requested: ");
+        }
+        
+        double HardwareMonitor::getAverageHotSpotTemperature()
```

Comment:
Can we unify 3 getAverage functions?
I think we just need to pass the reference of vector (and error message if we would like to change).

Reply:
Same for getMedian

Reply:
Functions unified